📝 Add User Task Documentation (#13)
atlas-engine/ContributionGuidelines: docs/getting_started/bpmn_elements/user_task.md

# User Task

![Sample User Task](../images/sample_user_task.svg)

<img width="250px" src="../images/user_task_property-panel.png">

> A user task is a task that requires the attention of a user to complete.

Reviewer: Ein ManualTask erfordert auch die Aufmerksamkeit des Users.

Hier sollte kurz illustriert werden, worin sich ManualTask und UserTask unterscheiden.

Each user task can have several options for asking the user for information. These options are called Form Fields.

## Form Field

A Form Field represents information that the process requires from the user.

There are different types of form fields:

- `text`
- `integer`
- `decimal`
- `truth value`
- `date`
- `enumeration`
- `custom type`

Each Form Field has:
- an ID which is used in the token,
- a label which gets displayed for the user,

> - and a default value which gets used if the user does not change the value.

Reviewer: Ich würde hier noch 2 Sektionen einfügen - analog zur ServiceTask Dokumentation:

- How it works, mit einem kleinen Diagramm das zeigt, wie ein User Task in der AE abgearbeitet wird
- Ein Diagramm mit einem praktischen Anwendungsbeispiel, auch damit der Unterschied zum Manual Task deutlicher wird

Reviewer: Das Diagramm kannst du in unserer Team Ablage unter "Atlas Engine Diagramme" ablegen. Da gibts auch schon ein Service Task Diagramm, dass du als Vorlage nehmen kannst.

# User Task

![Sample User Task](../images/sample_user_task.svg)

<img width="250px" src="../images/user_task_property-panel.png">

A User Task is a task that requires the attention of a user to complete.

> Unlike Manual Tasks the purpose of User Tasks is to receive information from the user.

Reviewer: Hab das mal etwas umformuliert, um das an die Beschreibung des Manual Tasks anzugleichen.

```suggestion
A User Task is used to delegate a unit of work to a user.
Unlike a Manual Task, a User Task describes a type of work that requires the user to actually provide some information, before it can be continued. 

Use this whenever you require the user to interact with your process.

For example: filling out a registration form or answering a question. 
```

> Each User Task can have several options for asking the user for information. These options are called Form Fields.

Reviewer: ```suggestion
User Tasks provide several options for asking the user for information. These options are called `Form Fields`.
```

## Form Field

A Form Field represents information that the process requires from the user.

> There are different types of form fields:

Reviewer: ```suggestion
There are different types of Form Fields:
```

- `text`
- `integer`
- `decimal`
- `truth value`
- `date`
- `enumeration`

> - `custom type`

Reviewer: Hier vllt noch nen kurzes Statement, was genau sich dahinter verbirgt. 

Fachfremde Anwender wissen z.B. vllt. nicht auf Anhieb, was ein "Integer" ist, oder wie dieser sich von "Decimal" unterscheidet.

Author: Aber integer ist doch einfach das englische Wort für eine Ganzzahl. Ich wüsste jetzt nicht wie ich diese Begriffe noch genauer erklären soll, ausser vielleicht `enumeration` und `custom type`

Reviewer: Dann ein kurzer Satz zu den beiden.

Each Form Field has:
- an ID which is used in the token,
- a label which gets displayed for the user,

> - and a default value which gets used if the user does not change the value.

Reviewer: ```suggestion
- a default value to use, if the user does not provide any input
```

# User Task

![Sample User Task](../images/sample_user_task.svg)

<img width="250px" src="../images/user_task_property-panel.png">

A User Task is a task that requires the attention of a user to complete.
Unlike Manual Tasks the purpose of User Tasks is to receive information from the user.

Each User Task can have several options for asking the user for information. These options are called Form Fields.

## Form Field

A Form Field represents information that the process requires from the user.

There are different types of form fields:

- `text`
- `integer`
- `decimal`
- `truth value`
- `date`
- `enumeration`
- `custom type`

Each Form Field has:
- an ID which is used in the token,
- a label which gets displayed for the user,
- and a default value which gets used if the user does not change the value.

## How it works

![How it works](../images/user_task_how_it_works.svg)

> 1. The User Task Task creates an User Task entry in the database and then pauses execution.

Reviewer: ```suggestion
1. The User Task creates a User Task entry in the database and then pauses execution.
```

1. An Atlas Engine Client that lies in an external application periodically polls for available User Tasks.
1. The Atlas Engine Client receives the User Task and waits for the user to enter the required information.
1. The Atlas Engine Client then returns the result to the Atlas Engine's User Domain.
1. The User Domain stores the User Task's result in the database and marks it as finished.
1. The User Domain communicates the result to the User Task, which in turn now finishes execution, using the User Task's result.

## Example

The following example shows a simple newsletter registration process.

![Example User Task Process](../images/example_user_task_process.svg)

In this example there is a User Task with the ID `enter_email_task`.

> It is used to ask the user for an email address that should receive newsletters.

Reviewer: ```suggestion
It is used to ask the user for an email address for receiving the newsletter.
```

> To do so it contains a Form Field of type `text` with the ID `email`.

Reviewer: ```suggestion
For this purpose, it has a Form Field of type `text` with the ID `email`.
```
